site-src/geps/gep-735.md

```diff
@@ -0,0 +1,178 @@
+request address matching for these APIs in order to support a standard
+for some of the commonplace setups of gateway implementations. Matching is
+intended to be covered for both _source_ and _destination_ to enable a finer
+level of tuning options for L4 traffic routing at a level below the `Gateway`.
+
+## API
+
+The API changes include the following new types:
+
+- `AddressMatch` to indicate the IP for address matching
+- `TrafficMatches` to configure matching according to network address
+
+These types enable the address matching required, with some active
+considerations about how to leave these open ended for later expansion.
+
+### AddressMatch Type
+
+A new `AddressMatch` type provides the targeting mechanism for match inclusion
+of a given network address:
+
+```go
+type AddressMatch struct {
+	// Type of the address, either IPAddress or NamedAddress.
+	//
+	// If NamedAddress is used this is a custom and specific value for each
+	// implementation to handle (and add validation for) according to their
+	// own needs.
+	//
+	// For IPAddress the implementor may expect either IPv4 or IPv6.
+	//
+	// Support: Core
+	//
+	// +optional
+	// +kubebuilder:validation:Enum=IPAddress;NamedAddress
+	// +kubebuilder:default=IPAddress
+	Type *AddressType `json:"type,omitempty"`
+
+	// Value of the address. The validity of the values will depend
+	// on the type and support by the controller.
+	//
+	// Examples: `1.2.3.4`, `128::1`, `my-named-address`.
+	//
+	// Support: Core
+	//
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	Value string `json:"value"`
+}
+```
+
+### TrafficMatches Type
+
+Using the new `AddressMatch` type matches can be expressed in topical lists on
+`TCPRoute` and `UDPRoute` using the new `TrafficMatches` type:
+
+```go
+type TrafficMatches struct {
+	// SourceAddresses indicates the originating (source) network
+	// addresses which are valid for routing traffic.
+	//
+	// Support: Core
+	SourceAddresses []AddressMatch `json:"sourceAddresses"`
+
+	// DestinationAddresses indicates the destination network addresses
+	// which are valid for routing traffic.
+	//
+	// Support: Core
+	DestinationAddresses []AddressMatch `json:"destinationAddresses"`
+```
+
+This type becomes an optional field and shared by both `TCPRouteRule` and
+`UDPRouteRule`:
+
+```go
+type TCPRouteRule struct {
+	// Matches add rules for filtering traffic to backends based on addresses.
+	//
+	// +optional
+	Matches *TrafficMatches `json:"matches"`
+}
+```
+
+The above would make the following YAML examples possible:
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1alpha2
+kind: TCPRoute
+metadata:
+  name: matching
+spec:
+  rules:
+  - matches:
+      sourceAddresses:
+      - value: "192.168.1.1"
+      - value: "FE80::0202:B3FF:FE1E:8329"
+      - type: NamedAddress
+        value: "my-custom-name"
+      destinationAddresses:
+      - value: "10.96.0.1"
+    backendRefs:
+    - name: my-service
+      port: 8080
```

Comment:
Any commentary on constitutes as the source/destination address when there are multiple hopes in the middle? For instance, is proxy protocol respected or not?

Reply:
Proxy protocol is an important point! I feel like I should send you a can opener for all the cans of worms you keep opening up. (It's a good thing!)

Personally, I feel like proxy protocol should really mean that the Gateway should be transparent and do minimal routing. So in the case that a Gateway has proxy protocol configured, that should mean that address-match routing like in this GEP should be disabled or ignored.

Reply:
It's funny you mention proxy protocol because I had a feeling this was going to come up. I'm somewhat inclined to add documentation that states we aren't considering it at this time and allow future contributors to expand into that and change that as needed. Does that sound like a reasonable course of action, or what would be your preference?

Reply:
I think that saying that proxy protocol support is undefined for now, PRs welcomed, here's a tracking issue, would make sense to get this functionality in.

Reply:
Sounds good, @hbagdi does this sound OK to you as a resolution?

Reply:
Proxy protocol being undefined _could_ break implementations in the future when we do define it.
For example, consider an implementation has decided to read proxy headers from the previous hop and takes that into account for routing. Now, if the spec clarifies later on that this must not be the case - that's a breaking change.
Is this understanding correct?


A safer approach would be to explicitly state if proxy protocol should be respected for routing decisions or not.

Reply:
I am persuaded, I have made an explicit point in the GEP about proxy protocol connection meta-data needing to be respected for match inclusions: 3ac0601fbbb1cb4b10f4b93e4b4f463f046c6130

Reply:
Given Harry's new commit https://github.com/kubernetes-sigs/gateway-api/pull/932/commits/57f90a8b3a22bbf69155c2f42ec13eb16258a0f4 it seems like we can resolve this comment now.